<Table> - add infiniteBulkSelection

splintor opened the pull request:
### 🔦 Summary
add infinite bulk selection. Enable a way to make `bulk selection` and `infiniteScroll` work together.
When a table has `infiniteScroll` and is not fully loaded, and the user does bulk selection ("Select All"), if the `infiniteBulkSelection` prop is set, then the `<Table>` considers all of the data rows as selected, including those not loaded yet, and excluded those that the user unselected after selecting all.
This is done by making `<BulkSelection>` hold either a `selectedIds` or `notSelectedIds` array, and when `selectAll()` is called, if the list is not fully loaded, switch to use `notSelectedIds` instead of `selectedIds`.

I still need to add a story to demonstrate this.
I don't think a visual test is needed for this, as the change is mainly in the behavior and not in the UI.

### ✅ Checklist
- [X] 👨‍💻 API change is approved by the UI Infra Developers @nitayneeman 
- [X] 👨‍🎨 UX change is approved by the Design System UX @shlomitc 
- 📚 Change is documented
  - [ ] Story
  - [X] API description
- 🔬 Change is tested
  - [X] Component

splintor commented on src/DataTable/DataTable.js:
```diff
@@ -154,8 +155,13 @@ class DataTable extends React.Component {
     } = this.props;
-    const rowClasses = [this.props.rowClass];
+    const rowClasses = [rowClass];
+    const key = defaultTo(rowData.id, rowNum);
```
I also prefer the vanilla way, but `<Table>` used `defaultTo(id, index)`.
So I wanted `<DataTable>` to be consistent with it. Do you want me to change `<Table>` to use the vanilla way as well?
Note that this might break existing code, as it will start treating `0` and `''` as non-existing `id`.

splintor commented on src/Table/BulkSelection/BulkSelection.js:
```diff
@@ -61,12 +65,32 @@ export class BulkSelection extends React.Component {
+      this.state.notSelectedIds &&
+      !this.canBeInInfiniteBulkSelection(nextProps)
+    ) {
+      // cancel infiniteBulkSelection no longer relebant (e.g. when done loading)
```
Oops. Fixed, thanks.

splintor commented on src/Table/components/TableContent.js:
```diff
@@ -27,7 +28,11 @@ export const TableContent = ({ titleBarVisible, dataHook }) => {
-                  selectedRowsIds={bulkSelectionContext.getSelectedIds()}
+                  isRowSelected={(rowData, rowIndex) =>
+                    bulkSelectionContext.isSelected(
+                      defaultTo(rowData.id, rowIndex),
```
Same answer - wanted to be consistent with existing behavior. Let me know if you want me to change it.

splintor commented on src/DataTable/DataTable.js:
```diff
@@ -584,7 +592,7 @@ DataTable.propTypes = {
   /** ++EXPERIMENTAL++ Set virtualized table row height */
   virtualizedLineHeight: PropTypes.number,
-  /** array of selected ids in the table */
+  /** array of selected ids in the table. This property is deprecated. Use `isRowSelected` instead. */
```
👍 done

splintor commented on src/Table/BulkSelection/BulkSelection.js:
```diff
@@ -85,39 +109,79 @@ export class BulkSelection extends React.Component {
+    }
   };
 
+  canBeInInfiniteBulkSelection(props) {
```
👍 Great idea! I was looking for the right name for this function and couldn't came up with a good idea. Thanks 🙏

splintor commented on src/Table/BulkSelection/BulkSelection.js:
```diff
@@ -61,12 +65,32 @@ export class BulkSelection extends React.Component {
       !this.areSelectedIdsEqual(nextProps.selectedIds, this.state.selectedIds)
     ) {
       this.setSelectedIds(nextProps.selectedIds.slice(), undefined, nextProps);
+    } else if (
```
The first `if` handles cases where the `selectedIds` prop changes (and then we copy the new `selectedIds` array we got and reset selection)
The second `if` handles cases where `allIds` prop changes. This happens in `infiniteScroll` when the user scrolls down. If the user previously did bulk selection and `bulkSelectionState` is `ALL`, and we are not in `infiniteBulkSelected`, then when the user scrolls down and more items are fetched (and `allIds` is updated), we want `bulkSelectionState` to change to `SOME` because not we have some new, unselected items, but selection doesn't need to change.

I hope this makes it clearer. I also have a specific unit test for that - `should update bulkSelectionState after loading more when not in infiniteBulkSelection`

splintor commented on src/Table/BulkSelection/BulkSelection.js:
```diff
@@ -128,25 +192,42 @@ export class BulkSelection extends React.Component {
+    const selectedCount = selectedIds
+      ? selectedIds.length
+      : totalCount - notSelectedIds.length;
+    const bulkSelectionState = selectedIds
```
Agree. Done.

splintor commented on src/Table/BulkSelection/BulkSelection.js:
```diff
@@ -128,25 +192,42 @@ export class BulkSelection extends React.Component {
       /** Is the item with the given id selected. (id comes from the rowData.id if exists, if not then it is the rowIndex) */
-      isSelected: id => selectedIds.indexOf(id) !== -1,
+      isSelected: id =>
+        selectedIds ? selectedIds.includes(id) : !notSelectedIds.includes(id),
```
It doesn't include a **selection id**. It includes **an id**. The `isSelected` function determines if a given `id` is selected. If `selectedIds` is defined, it checks if it is included in `selectedIds`. If not, it means that `notSelectedIds` is defined and then it checks if the given `id` is **not** included in `notSelectedIds`.

Do you think this code need to be changed to become clearer? If so, any idea on how to do it?

nitayneeman replied:
Perheps improving the comment above and emphasizing that `selectedIds` and `notSelectedIds` cannot be defined at once.

splintor replied:
I added an internal comment (as this is an implementation detail that doesn't belong in **Context API**)

splintor commented on src/Table/BulkSelection/BulkSelection.js:
```diff
@@ -188,10 +269,20 @@ BulkSelection.propTypes = {
+  /** The table's infiniteBulkSelection prop  */
+  infiniteBulkSelection: bool,
+  /** The table's hasMore prop  */
+  hasMore: bool,
```
Great Idea!
done

splintor commented on src/Table/Table.js:
```diff
@@ -128,14 +139,18 @@ export class Table extends React.Component {
+            disabled={selectionDisabled}
+            infiniteScroll={infiniteScroll}
+            infiniteBulkSelection={infiniteBulkSelection}
+            hasMore={hasMore}
```
Indeed. Done.

nitayneeman replied:
For a reason, I still see we pass these props:
https://github.com/wix/wix-style-react/blob/487cd2562217b6b63fdf946fe14b31f1c5688b13/src/Table/Table.js#L143

splintor replied:
Oops, you are right. I change `<BulkSelection>` props, but somehow my change in the place where it is being used in `Table.js` was reverted. Fixed. Thanks!

splintor commented on src/Table/BulkSelection/BulkSelection.js:
```diff
@@ -29,6 +29,8 @@ export const BulkSelectionContextPropTypes = {
   selectedCount: number,
   getSelectedIds: func,
+  getNotSelectedIds: func,
+  infiniteBulkSelected: bool,
```
I don't think so.
`infiniteBulkSelection` is a prop passed **by the developer** that indicates whether we should use infinite bulk selection behavior.
`infiniteBulkSelected` is a prop we pass **to the developer** that indicates whether the user has clicked **bulk selection** while we are in infinite bulk selection.

splintor commented on src/Table/docs/examples/TableInfiniteScrollExample.js:
```diff
@@ -1,46 +1,80 @@
 /* eslint-disable */
```
Agreed. I split the examples.

splintor commented on src/Table/docs/examples/TableInfiniteScrollExample.js:
```diff
@@ -1,46 +1,80 @@
     this.setState({ container: this.containerRef });
   }
 
+  renderToolbar = selectionContext => {
```
Not sure what you mean. I removed `showSelection` from state, as now in this sample it is always true, but we still need a toolbar in that example to show the user the number of selected items and to have a checkbox that sets the `infiniteBulkSelection` prop.

nitayneeman replied:
Right, I wasn't clear, we still need the toolbar for displaying the selection counter and checkbox.

nitayneeman commented:
We've some conflicts in the CHANGELOG (that's why we usually modify it after the PR is merged)

splintor commented:
I resolved the conflict.
How do we proceed?

nitayneeman commented:
> I resolved the conflict.
> How do we proceed?

Could be a bug with the example?
I remember it worked correctly.

![demo](https://user-images.githubusercontent.com/7141972/64528468-41999600-d311-11e9-86bc-3ec263a650b7.gif)

splintor commented:
The prop problem you found is the reason the example didn't work. Now that I fixed it, it should be OK.
I also added a commit with fixes to the documentation. I just realized that the `Context API` in the story is not taken from `BulkSelection.js` but has its own `README` file that needs to be updated. (also fixed some minor bugs there, like replacing `<` with `&lt;`)

nitayneeman commented:
@giladsegal and I thought about applying this behvior by default.

So, in order to avoid a breaking change - we'll provide an `upgrade` prop instead of `infiniteBulkSelection`. This means that in case we proceed with this thought, we actually can remove `infiniteBulkSelection`.

When I asked Ayelet about making it as the default behavior - I remember she encouraged that, but I will verify once again.

@shlomitc What do you think?

splintor commented:
@nitayneeman keep in mind that "by default", `<Table>` has no infinite behavior. Users need to opt-in for infinite scroll.
So, maybe, instead of having to specify both `infiniteScroll` and `infiniteBulkSelection` props, we can think of another way:
1. replace `infiniteScroll` with `infiniteBehavior` which can have values of `simple` and `bulkSelection` and `bulkSelection` would be the default.

2. Make `infiniteScroll` and `infiniteBulkSelection` mutual exclusive, so either one can be specified, and the developer needs to choose the selection behavior she wants.

What do you think?

giladsegal commented on src/DataTable/DataTable.js:
```diff
@@ -2,6 +2,7 @@ import React, { Component } from 'react';
 import PropTypes from 'prop-types';
 import styles from './DataTable.scss';
 import classNames from 'classnames';
+import defaultTo from 'lodash/defaultTo';
```
We usually try to avoid using `lodash`. Importing a single operator can have a surprising effect on the bundle size. We try to use it only for complex operators. You can just use rowData.id || rowNum

splintor replied:
I'd love to!

I already [got this comment from Nitay](https://github.com/wix/wix-style-react/pull/4046#discussion_r320677840).

As I wrote there, `<Table>` is [already](https://github.com/wix/wix-style-react/blob/master/src/Table/Table.js#L35) [using](https://github.com/wix/wix-style-react/blob/master/src/Table/Table.js#L115) `defaultTo` to get row id, so I thought the logic should be consistent with `<DataTable>`. If you think it is OK, I'd love to change that to use `rowData.id || rowNum`, but will it break existing code?

nitayneeman commented on src/Table/Table.js:
```diff
@@ -121,14 +132,18 @@ export class Table extends React.Component {
+            deselectRowsByDefault={deselectRowsByDefault}
+            disabled={selectionDisabled}
+            hasMoreInBulkSelection={
+              infiniteScroll && infiniteBulkSelection && hasMore
```
I talked to @shlomitc and he suggested to use the `totalSelectableCount` prop as an indicator for the bulk selection (instead of `infiniteBulkSelected`). I agree with him and think this thought would work because of the fact that `infiniteBulkSelected` doesn't have meaning without `totalSelectableCount` anyway...

What do you think?

splintor replied:
`infiniteBulkSelection` does have meaning without it. It means that when more items are loaded after the user did a bulk selection, the loaded items will be selected by default.

Using `totalSelectableCount` has no connection to this behavior. Users might set this property even without have an infinite scroll, or without knowing that setting it will cause loaded items to be selected.

nitayneeman replied:
>Users might set this property even without have an infinite scroll

I think this situation could happen also with the current implementation. In this case, it should do nothing IMO (maybe in the future it will, but not at the moment).

>or without knowing that setting it will cause loaded items to be selected

That's related to awareness. I believe that pretty well docs might solve this problem. In practice it means to move the relevant parts from `infiniteBulkSelection`'s API docs to `totalSelectableCount`.

So the suggestion is just changing the prerequisite of `infiniteBulkSelection` to `infiniteScroll` when using `totalSelectableCount`.

I've to admit that I see with this suggestion two concerns:
1. `totalSelectableCount` has a couple of responsibilities (indicator for the infinite bulk selection and also holds the number). Indeed, it might complicate the code and make it a bit nontrivial.
2. In case we would want to allow the consumers to perform an infinite bulk selection that doesn't select un-loaded items by default, for a reason, we'll apparently need to separate the indicator responsibility above from this prop.

But maybe it's worth the trade-off, when knowing that we don't supply a prop we need to maintain probably forever?

splintor replied:
> 2. In case we would want to allow the consumers to perform an infinite bulk selection that doesn't select un-loaded items by default
I don't understand. The main idea of "infinite bulk selection" is to treat not-yet-loaded items as selected. If they are not being selected when loaded, we can keep the behavior as it is today, but simply not specifying `totalSelectableCount`. What am I missing?

But there is another thing I didn't think of: What if users don't know the total selectable count? They want newly loaded items to be selected, but don't know the full amount of items and don't need to show it? (we had to ask MSS for a count API especially for that. Maybe not everyone has it)

nitayneeman replied:
> > 1. In case we would want to allow the consumers to perform an infinite bulk selection that doesn't select un-loaded items by default
> 
> I don't understand. The main idea of "infinite bulk selection" is to treat not-yet-loaded items as selected. If they are not being selected when loaded, we can keep the behavior as it is today, but simply not specifying `totalSelectableCount`. What am I missing?
> 

I agree with you, this is the main idea. I don't think we will implement this behavior, but rather just was an example for edge-case we'll need another prop. However, we still need `totalSelectableCount` to display the total selectable items counter.

> But there is another thing I didn't think of: What if users don't know the total selectable count? They want newly loaded items to be selected, but don't know the full amount of items and don't need to show it? (we had to ask MSS for a count API especially for that. Maybe not everyone has it)

As I said, we have to get this number in order to display the total count. Maybe we need to improve the prop to support an async function either, but `totalSelectableCount` is a constraint (otherwise, how would you know which number to render?).  

Actually I considered these capabilities as one single feature that is aimed only when using the infinite scroll, so that:
1. It makes the counter to consider all items (including those aren't rendered)
2. It selects all rendered items
3. When loading more items, it selects them by default

These are the requirements I understood from the feature at the beginning, meaning that the first and the third arrive together.

Regarding allowing only the third regardless infinite scroll - perhaps, but we need to check it with UX. I'd not do treat that as part of the scope of this PR. Also, in case we go just with `totalSelectableCount`, we always can detach the logic into another prop like `selectLoadedItems` or whatever.

splintor replied:
The main feature is the third one and also:
4. Be able to report what items are **not selected** when user did bulk selection.

1 is a nice-to-have, and it is also does not have to be part of the component. After all, the developer that uses is in charge of adding the `<SelectedCount>` where is it needed, and set its content, and if the developer knows the total selectable count, she can eaily do the math on her own, so it looks a bit strange to me to make this the main prop that triggers the `3` and `4` behaviors.

> how would you know which number to render?
Obviously, if we don't know the number, we won't show it. We will just know that all items are selected, and do something with it, but won't show the user the number of selected items. As I said, this is not the main feature.

Actually, we can change `SelectionContext.selectedCount` to `SelectionContext.getSelectedCount(totalSelectableCount)`, so we don't even need it as a prop...

nitayneeman replied:
> The main feature is the third one and also:
> 4. Be able to report what items are **not selected** when the user did bulk selection.

Why is this capability so important? What are you going to do with this data?

> `1` is a nice-to-have, and it also does not have to be part of the component. After all, the developer that uses is in charge of adding the `<SelectedCount>` where is it needed, and set its content, and if the developer knows the total selectable count, she can easily do the math on her own, so it looks a bit strange to me to make this the main prop that triggers the `3` and `4` behaviors.

I'm not sure it's a nice-to-have feature (at least I didn't know that originally). I think it should be verified from the UX perspective. Anyway it would be nice to provide this behavior for the consumer instead of implementing by his own, isn't it?

> > how would you know which number to render?
> > Obviously, if we don't know the number, we won't show it. We will just know that all items are selected, and do something with it, but won't show the user the number of selected items. As I said, this is not the main feature.
> 
> Actually, we can change `SelectionContext.selectedCount` to `SelectionContext.getSelectedCount(totalSelectableCount)`, so we don't even need it as a prop...

By saying "prop" do you mean that `SelectionContext.selectedCount` is redundant? If so, we can (in a separate PR of course) - but it sounds like a breaking change...

Well, so to recap:
- From your side (SLFP), 3 & 4 that were mentioned above are the must-have features.
- We need to understand from UX what are their must-have requirements (in other words, sort of a UX review)

In case UX considers the capabilities above (1, 3, 4) together as part of the "infinite bulk selection" feature - I'd try to proceed with a single prop (we can discuss the name and if WSR does the counter calculation) - that's the only way I see in order to promote this PR and we need a UX review anyway...

splintor replied:
> Why is this capability so important? What are you going to do with this data?

In SLFP we want to enable the user to select all sites (in a folder, or in search results) and move them to another folder. We can and will show the number of selected sites, but we would get value from the ability to select all sites, even those not yet loaded, in order to move all of them. We send the request to the server, and the server is in charge of getting the full list of sites and move them, so not all of them need to be loaded to the client.

> Anyway it would be nice to provide this behavior for the consumer

I agree. As I said, it is **nice** to have...

Please note that the PR adds no UX related to the total count. It only properly updates the `SelectionContext.selectedCount` prop. It is the consumer's job to use that prop and display it where needed, so I don't see how UX is related to it.

> do you mean that `SelectionContext.selectedCount` is redundant?

No, I meant `totalSelectableCount` might be redundant and can be passed as a method parameter, but I forgot that `SelectionContext.selectedCount` existed before my PR and it is problematic to remove it, so forget this idea.

After clarifying my concerns, if you guys still think having only `totalSelectableCount` prop to control this feature is OK, I am are totally fine with it. It fully meets our SLFP needs. I was just worried about some hypothetical consumer (that will probably never exist), and by the clarity of the feature. If it is OK with you, I'll go ahead and implement it.

PS
This hypothetical user can always pass some random number or `Number.POSITIVE_INFINITY` as `totalSelectableCount` if she just wants `3` & `4` and doesn't care about `1` and plan to display selected count in some other way.

nitayneeman replied:
Fair enough, let's do that.

nitayneeman commented on CHANGELOG.md:
```diff
@@ -25,6 +25,7 @@ Types of changes:
 - `<FloatingNotification/>` - add `width` prop [#4031](https://github.com/wix/wix-style-react/pull/4031)
 - `Cheatshhet`- adding links and examples of the `FoundationFamily` [#4038](https://github.com/wix/wix-style-react/pull/4038)
 - `<Table/>` - create `Table.BulkSelectionCheckbox` component [#4032](https://github.com/wix/wix-style-react/pull/4032)
+- `<Table/>` - add `infiniteBulkSelection` prop to better deal with bulk selection in `infiniteScroll` [#4046](https://github.com/wix/wix-style-react/pull/4046)
```
Fix the description

nitayneeman commented on src/Table/Table.js:
```diff
@@ -243,6 +258,10 @@ Table.propTypes = {
   /** If true, table will not render all data to begin with, but will gradually render the data as the user scrolls */
   infiniteScroll: PropTypes.bool,
+  /** If true, when `infiniteScroll` is set, and user choose "Select All", selection context will hold excluded items instead of selected items, as not all items might be loaded */
+  infiniteBulkSelection: PropTypes.bool,
```
Remove (but take the comment above)

nitayneeman commented on src/Table/Table.js:
```diff
@@ -108,8 +107,20 @@ export class Table extends React.Component {
+      deselectRowsByDefault,
+      selectionDisabled,
+      infiniteScroll,
+      infiniteBulkSelection,
```
Remove

nitayneeman commented on src/Table/docs/examples/TableInfiniteScrollWithBulkSelectionExample.js:
```diff
@@ -0,0 +1,89 @@
+            <TableToolbar.SelectedCount>{`${selectionContext.selectedCount} Selected`}</TableToolbar.SelectedCount>
+          </TableToolbar.Item>
+        </TableToolbar.ItemGroup>
+        <TableToolbar.ItemGroup position="end">
```
Probably redundant?

nitayneeman commented on src/Table/docs/index.story.js:
```diff
@@ -177,10 +178,17 @@ export default {
+            {
+              title: 'Table with Infinite Scroll and Bulk Selection',
+              description:
+                'When `infiniteScroll` (see the previous example) and `showSelection` are set, and the data is not fully loaded (`hasMore` is true), and then the user clicks the bulk selection checkbox, by default only loaded items are selected. If `infiniteBulkSelection` is set, then when the bulk selection checkbox is clicked before all the data is loaded, the grid enters an "infinite bulk sele ...
```
Fix the description

ayeletr commented:
Let's talk about the "Loading..." at the end of the table scroll. I think we should do it better.

splintor commented:
@Ayelet Reuveni <ayeletr@wix.com> we implemented it according to
https://zpl.io/V0pELWE.

Please close it with @Ronnie Weiss <ronniew@wix.com> and we'll update.

On Sun, Sep 15, 2019 at 11:24 AM ayeletr <notifications@github.com> wrote:

> Let's talk about the "Loading..." at the end of the table scroll. I think
> we should do it better.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/wix/wix-style-react/pull/4046?email_source=notifications&email_token=AAB3OXI4EFXFBHQZJ5QKRMDQJXWJ3A5CNFSM4ITHN7T2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6XL35Y#issuecomment-531545591>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAB3OXIR5ZFKFQERO2BGBQDQJXWJ3ANCNFSM4ITHN7TQ>
> .
>

ayeletr commented:
Yes :) But we will take it. It's not related to your task :)
